crates/sui-indexer/src/handlers/mod.rs

```diff
@@ -44,3 +52,112 @@ pub struct EpochToCommit {
+            .get_watermark_hi()
+            .await?
+            .map(|n| n.saturating_add(1))
+            .unwrap_or_default();
+
+        loop {
+            if cancel.is_cancelled() {
+                return Ok(());
+            }
+
+            // Try to fetch new data envelope from the stream
+            match stream.next().now_or_never() {
+                Some(Some(envelope_chunk)) => {
+                    if cancel.is_cancelled() {
+                        return Ok(());
+                    }
+                    for envelope in envelope_chunk {
+                        unprocessed.insert(envelope.sequence_number(), envelope);
+                    }
+                }
+                Some(None) => break, // Stream has ended
+                None => {}           // No new data envelope available right now
+            }
+
+            // Process unprocessed checkpoints, even no new checkpoints from stream
+            let checkpoint_lag_limiter = self.handler.get_checkpoint_lag_limiter().await?;
+            while next_cp_to_process <= checkpoint_lag_limiter {
+                if let Some(data_envelope) = unprocessed.remove(&next_cp_to_process) {
+                    envelope_batch.push(data_envelope);
+                    next_cp_to_process += 1;
+                } else {
+                    break;
+                }
+            }
+
+            if !envelope_batch.is_empty() {
+                let last_checkpoint_seq = envelope_batch.last().unwrap().sequence_number();
+                let batch = envelope_batch.into_iter().map(|c| c.data()).collect();
+                self.handler.load(batch).await.map_err(|e| {
+                    IndexerError::PostgresWriteError(format!(
+                        "Failed to load transformed data into DB for handler {}: {}",
+                        self.handler.name(),
+                        e
+                    ))
+                })?;
+                self.handler.set_watermark_hi(last_checkpoint_seq).await?;
+                envelope_batch = vec![];
+            }
+        }
+        Err(IndexerError::ChannelClosed(format!(
+            "Checkpoint channel is closed unexpectedly for handler {}",
+            self.handler.name()
+        )))
+    }
+}
+
+#[async_trait]
+pub trait Handler<T>: Send + Sync {
+    /// return handler name
+    fn name(&self) -> String;
+
+    /// commit batch of transformed data to DB
+    async fn load(&self, batch: Vec<T>) -> IndexerResult<()>;
+
+    /// read high watermark of the table DB
+    async fn get_watermark_hi(&self) -> IndexerResult<Option<u64>>;
+
+    /// set high watermark of the table DB, also update metrics.
+    async fn set_watermark_hi(&self, watermark_hi: u64) -> IndexerResult<()>;
+
+    /// get checkpoint lag limiter, for handlers without lag limiter, return u64::MAX
+    async fn get_checkpoint_lag_limiter(&self) -> IndexerResult<u64> {
+        Ok(u64::MAX)
+    }
+}
+
+pub trait CommitDataEnvelope<T>: Send + Sync {
```

Comment:
Seems unnecessary as a trait. Could just be a concrete struct with a pair of checkpoint sequence number and the data?

Reply:
makes sense, and replaced it with a tuple.